vhdx/src/vhdx_bat.rs

```diff
@@ -0,0 +1,126 @@
+pub const PAYLOAD_BLOCK_FULLY_PRESENT: u64 = 6;
+pub const PAYLOAD_BLOCK_PARTIALLY_PRESENT: u64 = 7;
+
+// mask for the BAT state
+pub const BAT_STATE_BIT_MASK: u64 = 0x07;
+// mask for the offset within the file in units of 1 MB
+pub const BAT_FILE_OFF_MASK: u64 = 0xFFFFFFFFFFF00000;
+
+macro_rules! div_round_up {
+    ($n:expr,$d:expr) => {{
+        ($n + $d - 1) / $d
+    }};
+}
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
+    // Invalid BAT entry
+    InvalidBatEntry,
+    // Invalid number of BAT entries
+    InvalidEntryCount,
+    // Failed to read BAT from file
+    ReadBat(io::Error),
+    // Failed to write BAT entries to file
+    WriteBat(io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBatEntry => write!(f, "Invalid BAT entry"),
+            InvalidEntryCount => write!(f, "Invalid BAT entry count"),
+            ReadBat(e) => write!(f, "Failed to read BAT entry {}", e),
+            WriteBat(e) => write!(f, "Failed to write BAT entry {}", e),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct BatEntry(pub u64);
+
+impl BatEntry {
+    // Read all BAT entries presented on the disk and insert them to a vector
+    pub fn collect_bat_entries(
+        f: &mut File,
+        disk_spec: DiskSpec,
+        bat_entry: RegionTableEntry,
+    ) -> Result<Vec<BatEntry>> {
+        let entry_count = BatEntry::calculate_entries(
+            disk_spec.block_size,
+            disk_spec.virtual_disk_size,
+            disk_spec.chunk_ratio,
+        );
+        if entry_count as usize > (bat_entry.length as usize / size_of::<BatEntry>()) {
+            return Err(Error::InvalidEntryCount);
+        }
+
+        let mut bat: Vec<BatEntry> = Vec::with_capacity(bat_entry.length as usize);
+        let offset = bat_entry.file_offset;
+        for i in 0..entry_count {
+            f.seek(SeekFrom::Start(offset + i * size_of::<u64>() as u64))
+                .map_err(Error::ReadBat)?;
+
+            let bat_entry = f.read_u64::<LittleEndian>().map_err(Error::ReadBat)?;
+            let bat_entry = BatEntry(bat_entry);
+            bat.insert(i as usize, bat_entry);
+        }
+
+        Ok(bat)
+    }
+
+    // Calculate the number of entries in the BAT
+    fn calculate_entries(block_size: u32, virtual_disk_size: u64, chunk_ratio: u64) -> u64 {
+        let data_blocks_count = div_round_up!(virtual_disk_size, block_size as u64);
+        data_blocks_count + (data_blocks_count - 1) / chunk_ratio
+    }
+
+    // Routine for writing BAT entries to the disk
+    pub fn write_bat_entries(
+        f: &mut File,
+        bat_offset: u64,
+        bat_entries: Vec<BatEntry>,
+    ) -> Result<()> {
+        let entry_count = bat_entries.len() as u64;
+        let offset = bat_offset;
```

Comment:
You don't need to create some intermediate variables here, just use them directly in the loop.